### PR TITLE
Fix migration for 3.3.0

### LINKED ISSUE #9709: Migrations failed when upgrading to 3.3
[//]: # ( Invisible comment: 
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Before you create the issue:
IIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIIII
Search for similar report among other reported issues.
Learn how to troubleshoot at https://www.mautic.org/docs/en/tips/troubleshooting.html
Use drag&drop to attach images or other files )

## Bug Description

When upgrading to 3.3 from 3.2.5 the migrations did not run correctly. The error was 

```
ALTER TABLE lead_lists ADD category_id INT UNSIGNED DEFAULT NULL -> ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categor
ies (id) ON DELETE SET NULL Migration 20210104171005 failed during Execution. Error An exception occurred while executing 'ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN K
EY (category_id) REFERENCES categories (id) ON DELETE SET NULL': \ SQLSTATE[HY000]: General error: 1005 Can't create table `#sql-965_c385` (errno: 150 "Foreign key constraint is incorrectly formed") 
```

The isse is, that in migration https://github.com/mautic/mautic/blob/3.3.0/app/migrations/Version20210104171005.php is run command 

```
ALTER TABLE {$this->prefix}lead_lists ADD category_id INT UNSIGNED DEFAULT NULL
```

Without explicitly set width of the column, in our case it created INT(10) and then the migration failed. After setting this field to INT(11) ran without an error.  

Safer to have 

```
ALTER TABLE {$this->prefix}lead_lists ADD category_id INT(11) UNSIGNED DEFAULT NULL
```

Probably the cause is the setting and version of Mysql server (5.5.5-10.2.16-MariaDB-10.2.16+maria~artful). 

Thanks

| Q   | A
| --- | ---
| Mautic version |  3.2.5
| PHP version | 7.4
| Mysql version | 5.5.5-10.2.16-MariaDB-10.2.16+maria~artful
| Browser | any

### Steps to reproduce
1. Upgrade to 3.3.0 

 
### Log errors

ALTER TABLE lead_lists ADD category_id INT UNSIGNED DEFAULT NULL -> ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categor
ies (id) ON DELETE SET NULL Migration 20210104171005 failed during Execution. Error An exception occurred while executing 'ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN K
EY (category_id) REFERENCES categories (id) ON DELETE SET NULL': \ SQLSTATE[HY000]: General error: 1005 Can't create table `#sql-965_c385` (errno: 150 "Foreign key constraint is incorrectly formed") 

[//]: # ( Invisible comment:
Please check for related errors in the latest log file in [mautic root]/app/log/ and/or the web server's logs and post them here. Be sure to remove sensitive information if applicable. )

### PR DESCRIPTION
**Pull request by alanhartless**
| Q                                      | A
| -------------------------------------- | ---
| Branch?                                | "3.3" 
| Bug fix?                               | no
| New feature?                           |no
| Deprecations?                          |no
| BC breaks?                             |no
| Automated tests included?              |no
| Related user documentation PR URL      | na
| Related developer documentation PR URL | na
| Issue(s) addressed                     | Fixes #9709

#### Description:
This addresses schema differences for M2 and M3 introduced in https://github.com/mautic/mautic/pull/9544. It should also remove the need for https://github.com/mautic/mautic/pull/9714 since it seems that some categories.id has a length of 10 and some 11 for different versions of MariaDB. 

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Preferably run the migrations on a failed upgrade, a M2  installation upgraded to M3, and an installation on M3. All three should work. (I've only tested on a M3 schema as I don't have a M2 schema but tried to manipulate the database to emulate all three scenarios).

### CODE REVIEW

**Review by RCheesley — DISMISSED** [review]
Updating from 2.16.5 to 3.2.4 and then manually to 3.3.0 at CLI resulted in the following error:

```
ruth.cheesley@MacBookPro-RuthCheesley local.mautic2 % php bin/console m:u:a --finish
Step    1 [->--------------------------] Migrating database schema...

An error occurred while updating the database. Check log for more details.
```
Applying your two file fixes as patches and then running `php bin/console doctrine:migrations:migrate` resulted in all the migrations being run correctly as expected.

php_version: "7.3"
mariadb_version: "10.2"

@dennisameling is also testing the PR in his local environment as a second test.

**Review by dennisameling — COMMENTED** [review]
Tried twice, didn't work...

- Installed Mautic 2 from ZIP package: `mkdir mautic2 && cd mautic2 && wget https://github.com/mautic/mautic/releases/download/2.16.5/2.16.5.zip && unzip -q 2.16.5.zip && ddev config --webserver-type=apache-fpm --php-version=7.3 && ddev start && ddev exec php app/console mautic:update:find` (ran the UI installer after this)
- Ran M2 > M3 upgrade: `php upgrade_v3.php` --> all good
- Ran `php bin/console m:u:f`
- Ran `php bin/console m:u:a`, this will attempt the 3.3.0 update. All good again!
- Ran `php bin/console m:u:a --finish`, error:

```
[2021-02-23 19:38:46] mautic.NOTICE: Doctrine\DBAL\Exception\DriverException: An exception occurred while executing 'ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categories (id) ON DELETE SET NULL':  SQLSTATE[HY000]: General error: 1005 Can't create table `db`.`lead_lists` (errno: 150 "Foreign key constraint is incorrectly formed") (uncaught exception) at /var/www/html/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php line 106 while running console command `doctrine:migrations:migrate` [] []
```

- Applied your commits:

```
wget -O patch-1.patch https://github.com/mautic/mautic/commit/ec770ab23c83f78446f9399bea7dcec3f2dc392f.patch
git apply patch-1.patch
wget -O patch-2.patch https://github.com/mautic/mautic/commit/5dadf7b6523b42c928f38ceaa17bec1341023df2.patch
git apply patch-2.patch
```

- Ran `php bin/console doctrine:migrations:migrate`
- Errors again:

```
WARNING! You are about to execute a database migration that could result in schema changes and data loss. Are you sure you wish to continue? (y/n)y
Migrating up to 20210223174702 from 20201207140911

  SS skipped (Reason: Schema includes this migration)

  ++ migrating 20210223174702

     -> ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categories (id) ON DELETE SET NULL
Migration 20210223174702 failed during Execution. Error An exception occurred while executing 'ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categories (id) ON DELETE SET NULL':

SQLSTATE[HY000]: General error: 1005 Can't create table `db`.`lead_lists` (errno: 150 "Foreign key constraint is incorrectly formed")

In AbstractMySQLDriver.php line 106:

  An exception occurred while executing 'ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categories (id) ON DELETE SET NULL
  ':

  SQLSTATE[HY000]: General error: 1005 Can't create table `db`.`lead_lists` (errno: 150 "Foreign key constraint is incorrectly formed")


In PDOConnection.php line 80:

  SQLSTATE[HY000]: General error: 1005 Can't create table `db`.`lead_lists` (errno: 150 "Foreign key constraint is incorrectly formed")


In PDOConnection.php line 75:

  SQLSTATE[HY000]: General error: 1005 Can't create table `db`.`lead_lists` (errno: 150 "Foreign key constraint is incorrectly formed")
```

**Comment by alanhartless** [comment]
Can you both get me the schema of your current categories and lead_lists tables?

**Comment by dennisameling** [comment]
Categories table:

image

image

Lead_lists table:

image

image

**Comment by RCheesley** [comment]
Categories

screenshot-mautic2 ddev site_8037-2021 02 23-21_08_01

Leads_list
screenshot-mautic2 ddev site_8037-2021 02 23-21_08_33

**Comment by dennisameling** [comment]
Interesting - my `category_id` in the `lead_lists` table is unsigned... That's probably causing the problem here

File: app/migrations/Version20210223174702.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        if (!$table->hasColumn('category_id')) {
+            $catTable         = $schema->getTable("{$this->prefix}categories");
+            $categoryIdColumn = $catTable->getColumn('id');
+            if ($categoryIdColumn->getUnsigned()) {
+                $this->addSql("ALTER TABLE {$this->prefix}lead_lists ADD category_id INT UNSIGNED DEFAULT NULL");
+            } else {
+                $categoryIdLength = $categoryIdColumn->getLength();
+                $this->addSql("ALTER TABLE {$this->prefix}lead_lists ADD category_id INT($categoryIdLength) DEFAULT NULL");
+            }
+        }
```
Comment (dennisameling) [inline app/migrations/Version20210223174702.php]:
```suggestion
        $catTable         = $schema->getTable("{$this->prefix}categories");
        $categoryIdColumn = $catTable->getColumn('id');

        // Add the new column if it failed for any reason
        if (!$table->hasColumn('category_id')) {
            if ($categoryIdColumn->getUnsigned()) {
                $this->addSql("ALTER TABLE {$this->prefix}lead_lists ADD category_id INT UNSIGNED DEFAULT NULL");
            } else {
                $this->addSql("ALTER TABLE {$this->prefix}lead_lists ADD category_id INT DEFAULT NULL");
            }
        } else {
            // The column was already added by the faulty migration, update it based on the type of the `category`.`id` column
            if ($categoryIdColumn->getUnsigned()) {
                $this->addSql("ALTER TABLE {$this->prefix}lead_lists CHANGE category_id category_id INT UNSIGNED DEFAULT NULL");
            } else {
                $this->addSql("ALTER TABLE {$this->prefix}lead_lists CHANGE category_id category_id INT DEFAULT NULL");
            }
        }
```

**Comment by dennisameling** [comment]
Okay, here's my assumption based on the screenshots:

- Prior to applying your PR, the migration already created the unsigned `category_id` column. The logic in the new migration, `20210223174702`, doesn't account for this scenario where the column was created but with the wrong type (unsigned), where it needs to be changed back from unsigned to regular first (or the other way around). Just added a suggestion to this PR to account for that scenario.

After applying that change, it works for me as well 🚀 

```
Migrating up to 20210223174702 from 20210104171005

  ++ migrating 20210223174702

     -> ALTER TABLE lead_lists CHANGE category_id category_id INT DEFAULT NULL
     -> ALTER TABLE lead_lists ADD CONSTRAINT FK_6EC1522A12469DE2 FOREIGN KEY (category_id) REFERENCES categories (id) ON DELETE SET NULL
     -> CREATE INDEX IDX_6EC1522A12469DE2 ON lead_lists (category_id)

  ++ migrated (took 5645.5ms, used 22M memory)

  ------------------------

  ++ finished in 5651ms
  ++ used 22M memory
  ++ 1 migrations executed
  ++ 3 sql queries
```

One thing I noticed is that `getLength()` returns `null`, causing the migration to fail. Is this method reliable?

image

**Comment by alanhartless** [comment]
Yes, that makes sense (fixing the faulting field created by the bad migration)

**Comment by alanhartless** [comment]
On the length, good question. I assumed so but obviously doesn't seem to be the case. Let me see what it comes up with for MySQL.

**Comment by alanhartless** [comment]
This PR seems confused now as it is including other commits despite rebasing on features. Cherry picked the commits into a new branch based on features and opened https://github.com/mautic/mautic/pull/9720

**Pull request closed without merging**